### DISCUSSION TITLE
When AOT Mapsui Sample no map is shown works in debug/release mode but not in AOT publish

### DISCUSSION THREAD

**Discussion by pilotshamn** [Mapsui/Mapsui#2317] [Q&A]
I am trying to publish a mapsui 4.1.1  blasor wasm app it works  when debuging and in publish without AOT mode but not in publish with AOT enable 

Then  I get this error
dotnet.7.0.14.ewp4hfjpf5.js:14  Uncaught RuntimeError: null function or function signature mismatch
    at dotnet.wasm:0x27cc93
    at dotnet.wasm:0x27ccc0
    at dotnet.wasm:0x27cd05
    at invoke_iii (dotnet.7.0.14.ewp4hfjpf5.js:14:258521)
    at dotnet.wasm:0x5daeb4
    at dotnet.wasm:0x1bd433
    at dotnet.wasm:0x45f46da
    at invoke_vi (dotnet.7.0.14.ewp4hfjpf5.js:14:260276)
    at dotnet.wasm:0x4610e65
    at dotnet.wasm:0x45f3cad

**Comment by pauldendulk** [comment]
We never looked into AOT and we definitely need to fix some things to make Mapsui compatible with AOT. I think many nuget package are not compatible with AOT and I am surprised people expect everything to work with AOT. I hope Microsoft thinks of a way to mark a nuget package compatible for AOT somehow (maybe there is already something for that), otherwise people will waste a lot on time on trial and error. 

Then again, it would be great if we could support AOT, especially because performance is important for us. Also, I don't think we use reflection in critical parts of the application. If would be nice if we could run an analyzer to tell us what is not compatible with AOT.

Things for us to read:
- https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7%2Cwindows
- https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows#aot-compatibility-analyzers

**Comment by pilotshamn** [comment]
Thank you for the feedback is there anything i can do? maybe brutile is the problem since skiasharp seem to function 
I have a quite complicated blazor component with a SKGLView and a lot of skiasharp in the code and it works in wasm AOT but no map in a mapsui component. If this can be of any help

**Reply by pauldendulk** [reply]
I created an issue to [investigate native AOT support](https://github.com/Mapsui/Mapsui/issues/2327). You could take a look. It is not an issue I want to dig into myself right now because there are so many other things to look into.

**Reply by inforithmics** [reply]
Without having looked to deep into it, I would target Blazor 8.0 for AOT, because a lot of AOT fixes went into the .NET 8 Framework. But for this I think the Mapsui.Blazor needs to add an .NET 8 target.

**Reply by inforithmics** [reply]
Pull Request to Add .NET 8 To Blazor in 4.1 Branch
https://github.com/Mapsui/Mapsui/pull/2329

**Reply by pauldendulk** [reply]
oh, that might simply be it. When going over the list of limitation I did not see any clear violations. They mention System.Reflection.Emit but we only use System.Reflection, so that should not be a problem.

**Reply by inforithmics** [reply]
Here pull request to enable AOT Validation and AOT Blazor Compilation in Main
https://github.com/Mapsui/Mapsui/pull/2349

**Comment by pilotshamn** [comment]
Hello
For what it is worth if I render a map dercetly to skiasharp it works in AOT se below
```csharp
bool showChart=true;
double chartResolution=8;
SKImageInfo  sKImageInfo;
static ITileSource baseTileSource;
TileLayer tileLayerMap;
Viewport vPort;
int chartType = 0;
Mapsui.Map map = new Mapsui.Map { CRS = "EPSG:3857" };
static MapControl mapControl = new MapControl();
MPoint centerOfRadarLonLat = new Mapsui.MPoint(SphericalMercator.FromLonLat(radarPosLon, radarPosLat).ToMPoint());
```

```csharp
protected override async Task OnInitializedAsync()
{
    sKImageInfo = new SKImageInfo(width, height);
    vPort = new Viewport
    {
        CenterX = centerOfRadarLonLat.X,
        CenterY = centerOfRadarLonLat.Y,
        Width = sKImageInfo.Width,
        Height = sKImageInfo.Height,
        Resolution = chartResolution       
    };
    chartType = await localStorage.GetItemAsync<int>("chartchoice");
    var attribution = new BruTile.Attribution("© OpenStreetMap contributors", "http://www.openstreetmap.org/copyright");
    var tileSource = await Verktyg.GetChoosenChart(chartType);
    tileLayerMap = new TileLayer(tileSource);
    map.Layers.Add(tileLayerMap);
    tileLayerMap.DataChanged += (s, e) =>
    {
        var layer = (TileLayer)s;
      if (!layer.Busy)
      {
         mapControl.Renderer.Render(chartSurface.Canvas, vPort, map.Layers, map.Widgets, col);
      }
    };
    mapControl.Map = map;
}

private void OnPaintSurface(SKPaintGLSurfaceEventArgs e) //SKGLView
{
    SKCanvas  canvas = e.Surface.Canvas;
    if (showChart)
    {
        fetchInfo = new FetchInfo(vPort.ToSection());
        tileLayerMap.RefreshData(fetchInfo);
    }
    if (showChart)
    canvas.DrawSurface(chartSurface, new SKPoint(0, 0));
}

```
